File: src/engine/detector.ts
// Feed-agnostic fee-adjusted arb detector. You feed it price ticks from ANY
// source (gRPC now, ShredStream later) and it emits arb open/close events with
// lifetime (slots + ms) = the reaction budget. A "real" arb requires the
// cross-venue gap to exceed the SUM of both venues' fees (net of fees).

export interface Tick {
  venue: string;
  price: number; // quote per base (e.g. USDC per SOL)
  slot: number;
  tsMs: number;
}

export interface ArbEvent {
  buyVenue: string;
  sellVenue: string;
  openSlot: number;
  closeSlot: number;
  lifetimeSlots: number;
  lifetimeMs: number;
  peakNetBps: number; // peak edge beyond both fees
}

export interface DetectorOpts {
  venues: [string, string];
  feeBps: Record<string, number>; // per-venue swap fee in bps
}

type OpenState = {
  buyVenue: string;
  sellVenue: string;
  openSlot: number;
  openTsMs: number;
  peakNetBps: number;
  lastSlot: number;
  lastTsMs: number;
};

export type TickResult =
  | { type: 'none'; netBps: number }
  | { type: 'open'; netBps: number }
  | { type: 'close'; event: ArbEvent };

export class Detector {
  readonly thresholdBps: number;
  private last: Record<string, number> = {};
  private open: OpenState | null = null;

  constructor(private opts: DetectorOpts) {
    const [a, b] = opts.venues;
    this.thresholdBps = (opts.feeBps[a] ?? 0) + (opts.feeBps[b] ?? 0);
  }

  onTick(t: Tick): TickResult {
    this.last[t.venue] = t.price;
    const [A, B] = this.opts.venues;
    const pa = this.last[A];
    const pb = this.last[B];
    if (pa === undefined || pb === undefined) return { type: 'none', netBps: 0 };

    const signedGapBps = ((pb - pa) / pa) * 10_000; // + => B dearer
    let net = 0;
    let buy = '';
    let sell = '';
    if (signedGapBps > this.thresholdBps) {
      net = signedGapBps - this.thresholdBps;
      buy = A;
      sell = B;
    } else if (signedGapBps < -this.thresholdBps) {
      net = -signedGapBps - this.thresholdBps;
      buy = B;
      sell = A;
    }

    if (net > 0) {
      if (!this.open) {
        this.open = {
          buyVenue: buy,
          sellVenue: sell,
          openSlot: t.slot,
          openTsMs: t.tsMs,
          peakNetBps: net,
          lastSlot: t.slot,
          lastTsMs: t.tsMs,
        };
        return { type: 'open', netBps: net };
      }
      this.open.peakNetBps = Math.max(this.open.peakNetBps, net);
      this.open.lastSlot = t.slot;
      this.open.lastTsMs = t.tsMs;
      return { type: 'none', netBps: net };
    }

    if (this.open) {
      const o = this.open;
      this.open = null;
      const closeSlot = t.slot || o.lastSlot;
      return {
        type: 'close',
        event: {
          buyVenue: o.buyVenue,
          sellVenue: o.sellVenue,
          openSlot: o.openSlot,
          closeSlot,
          lifetimeSlots: Math.max(0, closeSlot - o.openSlot),
          lifetimeMs: Math.max(0, t.tsMs - o.openTsMs),
          peakNetBps: o.peakNetBps,
        },
      };
    }
    return { type: 'none', netBps: 0 };
  }
}

const median = (a: number[]) => (a.length ? [...a].sort((x, y) => x - y)[Math.floor(a.length / 2)] : 0);

export function report(events: ArbEvent[], thresholdBps: number, runSec: number): void {
  console.log(`\n──────── shadow report (${runSec.toFixed(0)}s) ────────`);
  console.log(`Arb threshold (both fees): ${thresholdBps} bps`);
  console.log(`Real fee-adjusted arbs detected: ${events.length}`);
  if (!events.length) {
    console.log('  → none: pair arbed to within fees at this feed resolution.\n');
    return;
  }
  const nets = events.map((e) => e.peakNetBps);
  const slots = events.map((e) => e.lifetimeSlots);
  const ms = events.map((e) => e.lifetimeMs);
  const le1 = events.filter((e) => e.lifetimeSlots <= 1).length;
  console.log(`  peak net edge: median ${median(nets).toFixed(1)} bps, max ${Math.max(...nets).toFixed(1)} bps`);
  console.log(`  LIFETIME (reaction budget): median ${median(slots)} slots / ${median(ms)} ms, max ${Math.max(...slots)} slots`);
  console.log(`  closed within ≤1 slot: ${le1}/${events.length} (${((le1 / events.length) * 100).toFixed(0)}%) → need co-located/ShredStream speed`);
  console.log();
}

File: internal/detector/detector.go
// Package detector implements a feed-agnostic fee-adjusted arb detector — a
// direct port of the original TS/Rust version. Feed it price ticks from any
// source (gRPC now, ShredStream later) and it emits arb open/close with
// lifetime (slots + ms) = the reaction budget. A "real" arb requires the
// cross-venue gap to exceed the SUM of both fees.
package detector

import "sort"

type Tick struct {
	Venue string
	Price float64 // quote per base (USDC per SOL)
	Slot  uint64
	TsMs  uint64
}

type ArbEvent struct {
	BuyVenue      string
	SellVenue     string
	OpenSlot      uint64
	CloseSlot     uint64
	LifetimeSlots uint64
	LifetimeMs    uint64
	PeakNetBps    float64
}

type ResultKind int

const (
	ResultNone ResultKind = iota
	ResultOpen
	ResultClose
)

type TickResult struct {
	Kind   ResultKind
	NetBps float64  // valid when Kind == ResultOpen
	Event  ArbEvent // valid when Kind == ResultClose
}

type openState struct {
	buy, sell  string
	openSlot   uint64
	openTsMs   uint64
	peakNetBps float64
	lastSlot   uint64
	lastTsMs   uint64
}

type Detector struct {
	venueA, venueB string
	ThresholdBps   float64
	lastA, lastB   float64
	haveA, haveB   bool
	open           *openState
}

func New(a, b string, feeA, feeB float64) *Detector {
	return &Detector{venueA: a, venueB: b, ThresholdBps: feeA + feeB}
}

func (d *Detector) OnTick(t Tick) TickResult {
	if t.Venue == d.venueA {
		d.lastA, d.haveA = t.Price, true
	} else if t.Venue == d.venueB {
		d.lastB, d.haveB = t.Price, true
	}
	if !d.haveA || !d.haveB {
		return TickResult{Kind: ResultNone}
	}
	pa, pb := d.lastA, d.lastB

	signedGapBps := ((pb - pa) / pa) * 10000.0 // + => B dearer
	var net float64
	var buy, sell string
	switch {
	case signedGapBps > d.ThresholdBps:
		net, buy, sell = signedGapBps-d.ThresholdBps, d.venueA, d.venueB
	case signedGapBps < -d.ThresholdBps:
		net, buy, sell = -signedGapBps-d.ThresholdBps, d.venueB, d.venueA
	}

	if net > 0.0 {
		if d.open == nil {
			d.open = &openState{
				buy: buy, sell: sell,
				openSlot: t.Slot, openTsMs: t.TsMs,
				peakNetBps: net,
				lastSlot:   t.Slot, lastTsMs: t.TsMs,
			}
			return TickResult{Kind: ResultOpen, NetBps: net}
		}
		if net > d.open.peakNetBps {
			d.open.peakNetBps = net
		}
		d.open.lastSlot = t.Slot
		d.open.lastTsMs = t.TsMs
		return TickResult{Kind: ResultNone}
	}

	if d.open != nil {
		o := d.open
		d.open = nil
		closeSlot := t.Slot
		if closeSlot == 0 {
			closeSlot = o.lastSlot
		}
		lifetimeSlots := uint64(0)
		if closeSlot > o.openSlot {
			lifetimeSlots = closeSlot - o.openSlot
		}
		lifetimeMs := uint64(0)
		if t.TsMs > o.openTsMs {
			lifetimeMs = t.TsMs - o.openTsMs
		}
		return TickResult{Kind: ResultClose, Event: ArbEvent{
			BuyVenue: o.buy, SellVenue: o.sell,
			OpenSlot: o.openSlot, CloseSlot: closeSlot,
			LifetimeSlots: lifetimeSlots, LifetimeMs: lifetimeMs,
			PeakNetBps: o.peakNetBps,
		}}
	}
	return TickResult{Kind: ResultNone}
}

func MedianF64(v []float64) float64 {
	if len(v) == 0 {
		return 0.0
	}
	s := append([]float64{}, v...)
	sort.Float64s(s)
	return s[len(s)/2]
}

func MedianU64(v []uint64) uint64 {
	if len(v) == 0 {
		return 0
	}
	s := append([]uint64{}, v...)
	sort.Slice(s, func(i, j int) bool { return s[i] < s[j] })
	return s[len(s)/2]
}
